compiler/noirc_evaluator/src/ssa/opt/constant_folding/mod.rs

```diff
@@ -276,14 +338,27 @@ impl Context {
                     }
 
                     self.values_to_replace.batch_insert(&old_results, cached);
+
+                    // Because we removed some instruction from this block, we will have to (re)visit its successors.
+                    if origin != block {
+                        origins.reused.insert(origin);
+                    }
+
                     return;
                 }
-                CacheResult::NeedToHoistToCommonBlock(dominator) => {
+                CacheResult::NeedToHoistToCommonBlock { origin, dominator } => {
                     // Just change the block to insert in the common dominator instead.
                     // This will only move the current instance of the instruction right now.
                     // When constant folding is run a second time later on, it'll catch
                     // that the previous instance can be deduplicated to this instance.
+                    // Another effect is going to be that the cache should be updated to
+                    // point at the dominator, so subsequent blocks can use the result.
                     block = dominator;
+
+                    // Because we hoisted an instruction, we can revisit the origin to deduplicate.
+                    origins.hoisted.insert(origin);
```

Comment:
Perhaps I missed it, but I never see that `origins.hoisted` or `origins.reused` is cleared. Might be a cause of the memory regressions.

Reply:
It only lives for the duration of a block. 

I think the memory pressure comes from all the instructions and values to be replaced that keep accumulating during each revisit. Like if we have 10 instructions and replace 1, the subsequent 9 get new values and get inserted into the cache, then 8, 7, 6, ...